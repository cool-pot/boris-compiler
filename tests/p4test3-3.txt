*** test level 3, array local declare. local control flow
defun arrayfunc(x)
    local myarray;
    array myarray[5..10];
    local i = 5;
    local j = 5;
    local sum = 0;
    while i <= 10 do 
    	myarray[i] = 2*i;
    	i = i+1;
    end while
    while j <= 10 do 
    	sum = sum + myarray[j];
    	j = j+1;
    end while
    return sum * x;
end defun

print arrayfunc 3;