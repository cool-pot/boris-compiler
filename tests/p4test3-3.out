; ModuleID = 'module-boris-input'
source_filename = "module-boris-input"
target triple = "x86_64-pc-linux-gnu"

@format_int = private unnamed_addr constant [4 x i8] c"%d\0A\00"

define i32 @strlen(i8* %s) {
init:
  %i = alloca i32
  store i32 0, i32* %i
  br label %check

check:                                            ; preds = %body, %init
  %0 = load i32, i32* %i
  %1 = getelementptr i8, i8* %s, i32 %0
  %2 = load i8, i8* %1
  %3 = icmp ne i8 0, %2
  br i1 %3, label %body, label %end

body:                                             ; preds = %check
  %4 = load i32, i32* %i
  %5 = add i32 %4, 1
  store i32 %5, i32* %i
  br label %check

end:                                              ; preds = %check
  %6 = load i32, i32* %i
  ret i32 %6
}

declare i32 @printf(...)

define void @main() {
MainEntry:
  %0 = call i32 @arrayfunc(i32 3)
  %1 = call i32 (...) @printf(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @format_int, i32 0, i32 0), i32 %0)
  br label %EndEntry

EndEntry:                                         ; preds = %MainEntry
  ret void
}

define i32 @arrayfunc(i32) {
Entry:
  %x = alloca i32
  store i32 %0, i32* %x
  %myarray = alloca i32, i32 6
  %i = alloca i32
  store i32 5, i32* %i
  %j = alloca i32
  store i32 5, i32* %j
  %sum = alloca i32
  store i32 0, i32* %sum
  br label %while_init_

while_init_:                                      ; preds = %Entry
  br label %while_check_

while_check_:                                     ; preds = %while_body_, %while_init_
  %load_singleid_int_temp = load i32, i32* %i
  %1 = icmp sle i32 %load_singleid_int_temp, 10
  br i1 %1, label %while_body_, label %while_end_

while_body_:                                      ; preds = %while_check_
  %load_singleid_int_temp1 = load i32, i32* %i
  %2 = sub i32 %load_singleid_int_temp1, 5
  %3 = getelementptr i32, i32* %myarray, i32 %2
  %load_singleid_int_temp2 = load i32, i32* %i
  %expr_mul_expr_tmp = mul i32 2, %load_singleid_int_temp2
  store i32 %expr_mul_expr_tmp, i32* %3
  %load_singleid_int_temp3 = load i32, i32* %i
  %expr_plus_expr_tmp = add i32 %load_singleid_int_temp3, 1
  store i32 %expr_plus_expr_tmp, i32* %i
  br label %while_check_

while_end_:                                       ; preds = %while_check_
  br label %while_init_4

while_init_4:                                     ; preds = %while_end_
  br label %while_check_5

while_check_5:                                    ; preds = %while_body_6, %while_init_4
  %load_singleid_int_temp8 = load i32, i32* %j
  %4 = icmp sle i32 %load_singleid_int_temp8, 10
  br i1 %4, label %while_body_6, label %while_end_7

while_body_6:                                     ; preds = %while_check_5
  %load_singleid_int_temp9 = load i32, i32* %sum
  %load_singleid_int_temp10 = load i32, i32* %j
  %5 = sub i32 %load_singleid_int_temp10, 5
  %6 = getelementptr i32, i32* %myarray, i32 %5
  %load_array_ref_int_temp = load i32, i32* %6
  %expr_plus_expr_tmp11 = add i32 %load_singleid_int_temp9, %load_array_ref_int_temp
  store i32 %expr_plus_expr_tmp11, i32* %sum
  %load_singleid_int_temp12 = load i32, i32* %j
  %expr_plus_expr_tmp13 = add i32 %load_singleid_int_temp12, 1
  store i32 %expr_plus_expr_tmp13, i32* %j
  br label %while_check_5

while_end_7:                                      ; preds = %while_check_5
  %load_singleid_int_temp14 = load i32, i32* %sum
  %load_singleid_int_temp15 = load i32, i32* %x
  %expr_mul_expr_tmp16 = mul i32 %load_singleid_int_temp14, %load_singleid_int_temp15
  ret i32 %expr_mul_expr_tmp16
}
